### DISCUSSION TITLE
Why do route modules default to layout routes?

### DISCUSSION THREAD

**Discussion by cairin** [remix-run/remix#5930] [Q&A]
With both v1 and v2 route conventions, layout routes the default route module. 

Say I want two routes to be generated: `/pay` and `/pay/amount`. These are two distinct pages in an app, so it makes sense that they're separate route files, and neither of them are the layout of the other.

With the current route conventions I would simply have two files:

- `routes/pay.tsx`
- `routes/pay.amount.tsx`

With the current dot notation convention, the `pay.tsx` is not a layout route. However, if I replaced `routes/pay.amount.tsx` with `routes/pay/amount.tsx`, then `pay.tsx` is a layout route, so it follows that root files in any directory are layout routes.

Furthermore with v2 route conventions I now have:

- `routes/pay.tsx`
- `routes/pay_.amount.tsx`

This is to achieve the same flat route structure I had before, where neither is a layout of the other. 

If I have an app that has mostly distinct routes and very few layout routes. With v1 conventions most of my files just use dot delimiters, and this works quite nicely because it's easy to see my routes. However with v2 almost every filename needs to have trailing underscores (sometimes multiple times in a single filename) to ensure that they are not nested routes. In this case it would make the most sense to have layout routes with the trailing underscore, because there are much less of them.

Is it really that common that people need more layout routes than normal routes? If not (since v2 is a breaking change anyway) why not make normal routes the default?

**Comment by kiliman** [comment]
Can you give me an example where you would require multiple trailing underscores?

I typically add mine to the topmost parent that doesn't have an `<Outlet>`. Since flat-files can't use folders to define nested layouts, it uses a naming convention. Remix looks for the longest common prefix to find the parent layout. By using the `_`, you can ensure that a specific parent is *not* matched.

Everything is treated as nested here since we can always find a common prefix. 
```
routes/a.tsx
routes/a.b.tsx
routes/a.b.c.tsx
routes/a.b.c.d.tsx
```

To control the nesting, we add the `_` suffix. You don't have to add the `_` for *every* route that doesn't include an `<Outlet>`, just the topmost one, so Remix won't match that layout.

```
routes/a.tsx
routes/a.b.tsx
routes/a.b_.c.tsx  <- nest under a, not a.b
routes/a_.b.c.d.tsx <- nest under root
```

**Reply by cairin** [reply]
Using your example is probably the best way to demonstrate my concern. We have:
```
routes/a.tsx
routes/a.b.tsx
routes/a.b.c.tsx
routes/a.b.c.d.tsx
```
With route conventions v1 remix would generate somthing like this:
```html
<Routes>
  <Route file="root.tsx">
    <Route path="a" file="routes/a.tsx" />
    <Route path="a/b" file="routes/a.b.tsx" />
    <Route path="a/b/c" file="routes/a.b.c.tsx" />
    <Route path="a/b/c/d" file="routes/a.b.c.d.tsx" />
  </Route>
</Routes>
```
But, v2 looks more like this:
```html
<Routes>
  <Route file="root.tsx">
    <Route path="a" file="routes/a.tsx">
      <Route path="a/b" file="routes/a.b.tsx">
        <Route path="a/b/c" file="routes/a.b.c.tsx">
          <Route path="a/b/c/d" file="routes/a.b.c.d.tsx" />
        </Route>
      </Route>
    </Route>
  </Route>
</Routes>
```
For the sake of clarity let us assume there are no Outlets in these files except in `root.tsx`.

If a developer were to add an `<Outlet/>` to `routes/a.b.tsx`:
1. v1: Nothing changes for `routes/a.b.c.tsx` and `routes/a.b.c.d.tsx`. There are no side effects on these routes and that's clear when you look at your directory structure. The developer would need to explicitely refactor these files into a folder to make them nested.
2. v2: `routes/a.b.c.tsx` and `routes/a.b.c.d.tsx` would automatically be nested inside `routes/a.b.tsx`. However, in this case, there's nothing in the filename (`routes/a.b.tsx`) that tells me whether or not there is an `<Outlet/>` in the file. So the developer would need to explicitely opt out of file nesting using `_`.
 
The developer is happy with their changes and moves on - leaving the routes nested. A couple months later another developer comes along and adds the following routes:

```
routes/a.b.c.e.tsx
routes/a.b.c.f.tsx
routes/a.b.c.g.tsx
```

These routes are intended to be standalone and not nested under anything. However, the developer has no idea if there is an `<Outlet>` in any of the potential parent routes and now needs to check **all** the possible parent route files for possible nesting. 

Would it not be easier for the developer if the file with an `<Outlet>` was renamed to `routes/a.b_.tsx`. Essentially replacing directories in v1 with a trailing `_` in v2? In the example this would give the following route files:

```
routes/a.tsx
routes/a.b_.tsx <-- This is potentially a layout route (has an outlet).
routes/a.b_.c.tsx <-- I want this rendered in the outlet of b.
routes/a.b_.c.d.tsx <-- I also want this rendered in the outlet of b.
routes/a.b.c.e.tsx <-- I want this rendered in the outlet of root.
routes/a.b.c.f.tsx <-- I want this rendered in the outlet of root.
routes/a.b.c.g.tsx <-- I want this rendered in the outlet of root.
```

> Can you give me an example where you would require multiple trailing underscores?

Following the example above: if I wanted to ensure, without any shadow of a doubt, that `routes/a.b.c.g.tsx` will never render inside another parent layout route other than `root.tsx`, I would need to name it `routes/a_.b_.c_.g.tsx`. This is the only way to prevent it accidentally being rendered in one of its parent routes if they ever had an `<Outlet>` at some point in the future.

To summarise, layout routes have gone from opt-in in v1 to opt-out in v2.

**Reply by kiliman** [reply]
If you always want routes to be nested under root, make them all start with `a_`. Just think of the `_` as "_this is the start of my dotted routes of v1_".

Anyway, I'm not discounting your concern. But the spec has been out for almost a year. I've had a working version since last April. And Remix plans to release v2 any day now with this as the default. I am trying to explain how it works today.

BTW: You will still be able to opt-in to the v1 convention. Just like how you had to use `routes` config to opt-in to `remix-flat-routes`, you'll be able to do the same for `remix-legacy-routes` (not sure what the official name will be, though).

**Reply by kiliman** [reply]
P.S. Remix has created some compat utils

https://github.com/remix-run/v1-compat-utils/tree/main/packages/v1-route-convention

**Reply by cairin** [reply]
> If you always want routes to be nested under root, make them all start with a_. Just think of the _ as "this is the start of my dotted routes of v1".

Aha, this makes more sense. I must have misread your earlier comment because you did actually say this then. I somehow missed the fact that the **first** trailing underscore essentially ends further nesting at that point for that route. 

> Anyway, I'm not discounting your concern. But the spec has been out for almost a year. I've had a working version since last April. And Remix plans to release v2 any day now with this as the default. I am trying to explain how it works today.

Thanks, I did read the spec when it was initially set out, but today was the first time I used it when trying to migrate our app to use the new convention.

**Reply by cairin** [reply]
@kiliman They way you think this works, is not actually how it works.

> If you always want routes to be nested under root, make them all start with a_. Just think of the _ as "this is the start of my dotted routes of v1".

This statement is just not true.

**Reply by cairin** [reply]
If I am wanting to create the route `/a/b/c/d` and I already have the route `/a/b/c` which is created using the file `/a_.b.c.tsx` (so that it renders under root). In order for `/a/b/c/d` to also render exclusively under root I need to name my file `/a_.b.c_.d.tsx`. Otherwise it will just try render it in the outlet of `/a_.b.c.tsx` (which doesn't exist) and `/a_.b.c_.d.tsx` just won't render at all.

This happens for any file that already partially matches another file. So in order to deterministically know that my file naming is the same as v1 I need to put an `_` under every route to ensure it doesn't magically try nest things I don't want it to in the future.

**Reply by kiliman** [reply]
Hmm... I see. That's not typically a route I would generate, but I understand how that could happen. Perhaps the algorithm could be changed to never match a parent route that has a trailing underscore. At that point, once you add the underscore to `a`, it would always match with root. It's possible that was the intended behavior in the original spec without calling it out specifically.

I'll have to check my `remix-flat-routes` and see if it also has this issue. When I was implementing this, I remember figuring out how to nest layouts was the trickiest part.

**Reply by cairin** [reply]
> Perhaps the algorithm could be changed to never match a parent route that has a trailing underscore.

That's less configureable because you do sometimes want multiple nested layouts, and if I'm understanding you correctly - with this change that would not be possible.

I know it's late to the game in terms of Remix having already made this API stable for v2 etc. I'm only bringing this up now because this is the first I've used the API. But it feels like reversing the opt-in to nested layout (v1 implementation) to be opt-out (v2 implementation) is going to be horribly confusing for end users of remix.

To that end (as an end user) I feel like it would be less cognitive overhead to be able to opt in to a nested layout (like in v1). As you mentioned you need to replace directories with with a naming convention:

> Since flat-files can't use folders to define nested layouts, it uses a naming convention. Remix looks for the longest common prefix to find the parent layout. By using the `_`, you can ensure that a specific parent is _not_ matched.

Why not make the trailing `_` denote a specific parent _should_ match:

```
routes/a.tsx
routes/a.b_.tsx <- This is a layout route that I've place an Outlet in. Its parent is root.
routes/a.b_.c.tsx  <- nest under b as its immediate parent.
routes/a.b.c.d.tsx <- nest under root
```

With this notation I don't have any of the problems I've mentioned with the current implementation, because I know where the layout nesting is happeneing because I've specifically told remix to nest layouts there. And as an added bonus I can immediately see what routes possibly have an `<Outlet>` in them just by looking at my directory structure.

**Comment by ryanflorence** [comment]
I'd say keep using the v1 route convention if it feels better for your app: https://github.com/remix-run/v1-compat-utils/tree/main/packages/v1-route-convention

They're just conventions for folks without an opinion yet :)

**Reply by cairin** [reply]
I understand I can use whatever routing mechanism I please, but would prefer to use the built in one because it's less overhead.

The bigger problem here is that the [docs don't explain the current v2 convention correctly.](https://github.com/remix-run/remix/issues/5846#issuecomment-1543650188) Which is the main cause for the confusion here.
